packages/ics721/src/execute.rs

```diff
@@ -483,48 +500,52 @@ where
         create: VoucherCreation,
     ) -> Result<Response<T>, ContractError> {
         let VoucherCreation { class, tokens } = create;
-        let instantiate = if CLASS_ID_TO_NFT_CONTRACT.has(deps.storage, class.id.clone()) {
-            vec![]
-        } else {
-            let class_id = ClassId::new(class.id.clone());
-            let cw721_code_id = CW721_CODE_ID.load(deps.storage)?;
-            // for creating a predictable nft contract using, using instantiate2, we need: checksum, creator, and salt:
-            // - using class id as salt for instantiating nft contract guarantees a) predictable address and b) uniqueness
-            // for this salt must be of length 32 bytes, so we use sha256 to hash class id
-            let mut hasher = Sha256::new();
-            hasher.update(class_id.as_bytes());
-            let salt = hasher.finalize().to_vec();
-
-            let cw721_addr = get_instantiate2_address(
-                deps.as_ref(),
-                env.contract.address.as_str(),
-                &salt,
-                cw721_code_id,
-            )?;
-
-            // Save classId <-> contract mappings.
-            CLASS_ID_TO_NFT_CONTRACT.save(deps.storage, class_id.clone(), &cw721_addr)?;
-            NFT_CONTRACT_TO_CLASS_ID.save(deps.storage, cw721_addr, &class_id)?;
-
-            let admin = ADMIN_USED_FOR_CW721
-                .load(deps.storage)?
-                .map(|a| a.to_string());
-            let message = SubMsg::<T>::reply_on_success(
-                WasmMsg::Instantiate2 {
-                    admin,
-                    code_id: cw721_code_id,
-                    msg: self.init_msg(deps.as_ref(), &env, &class)?,
-                    funds: vec![],
-                    // Attempting to fit the class ID in the label field
-                    // can make this field too long which causes data
-                    // errors in the SDK.
-                    label: "ics-721 debt-voucher cw-721".to_string(),
-                    salt: salt.into(),
-                },
-                INSTANTIATE_CW721_REPLY_ID,
-            );
-            vec![message]
-        };
+        let instantiate =
+            if CLASS_ID_AND_NFT_CONTRACT_INFO.has(deps.storage, class.id.to_string().as_str()) {
+                vec![]
+            } else {
```

Comment:
This is a huge block doing a lot of calculations. Can we put this in a function somewhere?

Reply:
moved to `create_instantiate_msg()`

Reply:
Awesome